src/HamiltonianBuilder/DOCI.cpp

```diff
@@ -53,42 +53,80 @@ Eigen::MatrixXd DOCI::constructHamiltonian(const HamiltonianParameters& hamilton
+            // The e2 iteration counts the amount of encountered electrons for the creation operator
+            // We only consider greater addresses than the initial one (because of symmetry)
+            // Hence we are only required to start counting from the annihilated electron (e1)
+            size_t e2 = e1;
+
+            size_t q = p + 1;
+
+            // Test whether next orbital is occupied, until we reach unoccupied orbital
+            while (e2 < N - 1 && q == onv.get_occupied_index(e2 + 1)) {
+                // Shift the address for the electrons encountered after the annihilation but before the creation
+                // Their currents weights are no longer correct, the corresponding weights can be calculated
+                // initial weight can be found in the addressing scheme, on the index of the orbital (row) and electron count (column)
+                // since e2 starts at the annihilated position, the first shifted electron is at e2's position + 1, (given the while loop condition this is also (e2+1)'s position)
+                // The nature of the addressing scheme requires us the add 1 to the electron count (because we start with the 0'th electron
+                // And for the initial weight we are at an extra electron (before the annihilation) hence the difference in weight is:
+                // the new weight at (e2+1) position (row) and e2+1 (column) - the old weight at  (e2+1) position (row) and e2+2 (column)
+                address += this->fock_space.get_vertex_weights(q, e2 + 1) - this->fock_space.get_vertex_weights(q, e2 + 2);
+                e2++;  // adding occupied orbitals to the electron count
+                q++;
             }
-        }  // p or e1 loop
 
-        // Skip the last permutation
-        if (I < dim-1) {
+            e2++;
+            while (q < K) {
+                size_t J = address + this->fock_space.get_vertex_weights(q, e2);
+
+                // address has been calculated, since I is only updated in the outer loop
+                // it is more efficient to add it to a locally scoped variable and add it to the vector with one access.
+
+                result_matrix(I, J) += hamiltonian_parameters.get_g()(p, q, p, q);
+                result_matrix(J, I) += hamiltonian_parameters.get_g()(p, q, p, q);
+
+                // go to the next orbital
+                q++;
+
+                // if we encounter an occupied orbital, perform the shift, and test whether the following orbitals are occupied (or not)
+                // then proceed to set q to the next non-occupied orbital.
+                if (e2 < N && q == onv.get_occupied_index(e2)) {
+                    address += this->fock_space.get_vertex_weights(q, e2) - this->fock_space.get_vertex_weights(q, e2 + 1);
+                    q++;
+                    while (e2 < N - 1 &&  q == onv.get_occupied_index(e2 + 1)) {
+                        // see previous
```

Comment:
This is the second time you use this line of code: I think it's better to abstract it to a function, and write a corresponding unit test